# Analyzing CIA Factbook Data Using SQL

In this project, we'll work with data from the [CIA World Factbook](https://www.cia.gov/library/publications/the-world-factbook/), a compendium of statistics about all of the countries on Earth. The Factbook contains demographic information like:

- population - The population as of 2015.
- population_growth - The annual population growth rate, as a percentage.
- area - The total land and water area.

The SQLite database, factbook.db, is available in the repository.

We use the following code to connect our Jupyter Notebook to our database file:

In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

Note that we must use %%sql in every SQL cell, or else the code will be read as iPython.

## Overview of the Data

In [2]:
%%sql
SELECT *
FROM sqlite_master 
WHERE type='table';

 * sqlite:///factbook.db
Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

## Summarizing the Data

In [3]:
%%sql 
SELECT *
FROM facts 
LIMIT 5;

 * sqlite:///factbook.db
Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]


Here are the descriptions for some of the columns:

- name - The name of the country.
- area - The total land and sea area of the country.
- population - The country's population.
- population_growth- The country's population growth as a percentage.
- birth_rate - The country's birth rate, or the number of births a year per 1,000 people.
- death_rate - The country's death rate, or the number of death a year per 1,000 people.
- area- The country's total area (both land and water).
- area_land - The country's land area in square kilometers.
- area_water - The country's waterarea in square kilometers.

Let's start by calculating some summary statistics and look for any outlier countries.

## Identifying Outliers

In [4]:
%%sql
SELECT MIN(population) 'min_pop', MAX(population) 'max_pop', MIN(population_growth) 'min_pop_growth', MAX(population_growth) 'max_pop_growth'
FROM facts;

 * sqlite:///factbook.db
Done.


[(0, 7256490011, 0.0, 4.02)]

A few things stick out from the summary statistics:

- there's a country with a population of 0
- there's a country with a population of 7256490011 (or more than 7.2 billion people)

So we find the countries with 0 population and maximum population.

In [5]:
%%sql
SELECT name
FROM facts 
WHERE population = (SELECT MIN(population)
                   FROM facts);

 * sqlite:///factbook.db
Done.


[('Antarctica',)]

In [6]:
%%sql
SELECT name
FROM facts 
WHERE population = (SELECT MAX(population)
                   FROM facts);

 * sqlite:///factbook.db
Done.


[('World',)]

It seems like the table contains a row for the whole world, which explains the population of over 7.2 billion. It also seems like the table contains a row for Antarctica, which explains the population of 0.

We move on to calculating Averages now.

In [7]:
%%sql
SELECT ROUND(AVG(population), 2) 'avg_pop', ROUND(AVG(area), 2) 'avg_area'
FROM facts;

 * sqlite:///factbook.db
Done.


[(62094928.32, 555093.55)]

We will now identify the countries that have:
- above average values for population
- below average values for area


### Countries with population more than average population

In [8]:
%%sql
SELECT name
FROM facts
WHERE population > (SELECT AVG(population)
                   FROM facts);

 * sqlite:///factbook.db
Done.


[('Bangladesh',),
 ('Brazil',),
 ('China',),
 ('Congo, Democratic Republic of the',),
 ('Egypt',),
 ('Ethiopia',),
 ('France',),
 ('Germany',),
 ('India',),
 ('Indonesia',),
 ('Iran',),
 ('Japan',),
 ('Mexico',),
 ('Nigeria',),
 ('Pakistan',),
 ('Philippines',),
 ('Russia',),
 ('Thailand',),
 ('Turkey',),
 ('United Kingdom',),
 ('United States',),
 ('Vietnam',),
 ('European Union',),
 ('World',)]

### Countries with area less than average area

In [9]:
%%sql
SELECT name
FROM facts
WHERE area < (SELECT AVG(area)
                   FROM facts);

 * sqlite:///factbook.db
Done.


[('Albania',),
 ('Andorra',),
 ('Antigua and Barbuda',),
 ('Armenia',),
 ('Austria',),
 ('Azerbaijan',),
 ('Bahamas, The',),
 ('Bahrain',),
 ('Bangladesh',),
 ('Barbados',),
 ('Belarus',),
 ('Belgium',),
 ('Belize',),
 ('Benin',),
 ('Bhutan',),
 ('Bosnia and Herzegovina',),
 ('Brunei',),
 ('Bulgaria',),
 ('Burkina Faso',),
 ('Burundi',),
 ('Cambodia',),
 ('Cameroon',),
 ('Cabo Verde',),
 ('Comoros',),
 ('Congo, Republic of the',),
 ('Costa Rica',),
 ("Cote d'Ivoire",),
 ('Croatia',),
 ('Cuba',),
 ('Cyprus',),
 ('Czech Republic',),
 ('Denmark',),
 ('Djibouti',),
 ('Dominica',),
 ('Dominican Republic',),
 ('Ecuador',),
 ('El Salvador',),
 ('Equatorial Guinea',),
 ('Eritrea',),
 ('Estonia',),
 ('Fiji',),
 ('Finland',),
 ('Gabon',),
 ('Gambia, The',),
 ('Georgia',),
 ('Germany',),
 ('Ghana',),
 ('Greece',),
 ('Grenada',),
 ('Guatemala',),
 ('Guinea',),
 ('Guinea-Bissau',),
 ('Guyana',),
 ('Haiti',),
 ('Honduras',),
 ('Hungary',),
 ('Iceland',),
 ('Iraq',),
 ('Ireland',),
 ('Israel',),
 ('Italy',),
 ('Jamaica',),
 ('Japan',),
 ('Jordan',),
 ('Kiribati',),
 ('Korea, North',),
 ('Korea, South',),
 ('Kosovo',),
 ('Kuwait',),
 ('Kyrgyzstan',),
 ('Laos',),
 ('Latvia',),
 ('Lebanon',),
 ('Lesotho',),
 ('Liberia',),
 ('Liechtenstein',),
 ('Lithuania',),
 ('Luxembourg',),
 ('Macedonia',),
 ('Malawi',),
 ('Malaysia',),
 ('Maldives',),
 ('Malta',),
 ('Marshall Islands',),
 ('Mauritius',),
 ('Micronesia, Federated States of',),
 ('Moldova',),
 ('Monaco',),
 ('Montenegro',),
 ('Morocco',),
 ('Nauru',),
 ('Nepal',),
 ('Netherlands',),
 ('New Zealand',),
 ('Nicaragua',),
 ('Norway',),
 ('Oman',),
 ('Palau',),
 ('Panama',),
 ('Papua New Guinea',),
 ('Paraguay',),
 ('Philippines',),
 ('Poland',),
 ('Portugal',),
 ('Qatar',),
 ('Romania',),
 ('Rwanda',),
 ('Saint Kitts and Nevis',),
 ('Saint Lucia',),
 ('Saint Vincent and the Grenadines',),
 ('Samoa',),
 ('San Marino',),
 ('Sao Tome and Principe',),
 ('Senegal',),
 ('Serbia',),
 ('Seychelles',),
 ('Sierra Leone',),
 ('Singapore',),
 ('Slovakia',),
 ('Slovenia',),
 ('Solomon Islands',),
 ('Spain',),
 ('Sri Lanka',),
 ('Suriname',),
 ('Swaziland',),
 ('Sweden',),
 ('Switzerland',),
 ('Syria',),
 ('Tajikistan',),
 ('Thailand',),
 ('Timor-Leste',),
 ('Togo',),
 ('Tonga',),
 ('Trinidad and Tobago',),
 ('Tunisia',),
 ('Turkmenistan',),
 ('Tuvalu',),
 ('Uganda',),
 ('United Arab Emirates',),
 ('United Kingdom',),
 ('Uruguay',),
 ('Uzbekistan',),
 ('Vanuatu',),
 ('Holy See (Vatican City)',),
 ('Vietnam',),
 ('Yemen',),
 ('Zimbabwe',),
 ('Taiwan',),
 ('Ashmore and Cartier Islands',),
 ('Christmas Island',),
 ('Cocos (Keeling) Islands',),
 ('Coral Sea Islands',),
 ('Heard Island and McDonald Islands',),
 ('Norfolk Island',),
 ('Hong Kong',),
 ('Macau',),
 ('Faroe Islands',),
 ('Clipperton Island',),
 ('French Polynesia',),
 ('New Caledonia',),
 ('Saint Martin',),
 ('Saint Pierre and Miquelon',),
 ('Wallis and Futuna',),
 ('Aruba',),
 ('Curacao',),
 ('Sint Maarten',),
 ('Cook Islands',),
 ('Niue',),
 ('Tokelau',),
 ('Bouvet Island',),
 ('Jan Mayen',),
 ('Svalbard',),
 ('Akrotiri',),
 ('Anguilla',),
 ('Bermuda',),
 ('British Indian Ocean Territory',),
 ('British Virgin Islands',),
 ('Cayman Islands',),
 ('Dhekelia',),
 ('Falkland Islands (Islas Malvinas)',),
 ('Gibraltar',),
 ('Guernsey',),
 ('Jersey',),
 ('Isle of Man',),
 ('Montserrat',),
 ('Pitcairn Islands',),
 ('Saint Helena, Ascension, and Tristan da Cunha',),
 ('South Georgia and South Sandwich Islands',),
 ('Turks and Caicos Islands',),
 ('American Samoa',),
 ('Guam',),
 ('Navassa Island',),
 ('Northern Mariana Islands',),
 ('Puerto Rico',),
 ('Virgin Islands',),
 ('Wake Island',),
 ('Gaza Strip',),
 ('Paracel Islands',),
 ('Spratly Islands',),
 ('West Bank',),
 ('Western Sahara',)]

## Exploring water to land ratios

In [10]:
%%sql
SELECT name, CAST(area_water as float)/area_land 'ratio_w:l'
FROM facts;

 * sqlite:///factbook.db
Done.


[('Afghanistan', 0.0),
 ('Albania', 0.04927366961092051),
 ('Algeria', 0.0),
 ('Andorra', 0.0),
 ('Angola', 0.0),
 ('Antigua and Barbuda', 0.0),
 ('Argentina', 0.015971849204696183),
 ('Armenia', 0.054604120129064286),
 ('Australia', 0.007669578121135597),
 ('Austria', 0.017296379404451452),
 ('Azerbaijan', 0.048058187803313605),
 ('Bahamas, The', 0.3866133866133866),
 ('Bahrain', 0.0),
 ('Bangladesh', 0.1405085657217485),
 ('Barbados', 0.0),
 ('Belarus', 0.023164120256283883),
 ('Belgium', 0.008256820133430213),
 ('Belize', 0.00701569762343243),
 ('Benin', 0.018079586339064563),
 ('Bhutan', 0.0),
 ('Bolivia', 0.014105036365700761),
 ('Bosnia and Herzegovina', 0.0001953621036591322),
 ('Botswana', 0.026467630088401885),
 ('Brazil', 0.018859459161966657),
 ('Brunei', 0.0949667616334283),
 ('Bulgaria', 0.022029883213966393),
 ('Burkina Faso', 0.0014609203798392988),
 ('Burma', 0.03530178666519767),
 ('Burundi', 0.0837227414330218),
 ('Cambodia', 0.025606888932951873),
 ('Cameroon', 0.005775211017325633),
 ('Canada', 0.09799992456155804),
 ('Cabo Verde', 0.0),
 ('Central African Republic', 0.0),
 ('Chad', 0.019695044472681066),
 ('Chile', 0.01652299236903949),
 ('China', 0.029009018475490568),
 ('Colombia', 0.09647636468662751),
 ('Comoros', 0.0),
 ('Congo, Democratic Republic of the', 0.03432216697661452),
 ('Congo, Republic of the', 0.0014641288433382138),
 ('Costa Rica', 0.0007833920877399138),
 ("Cote d'Ivoire", 0.014025024921148543),
 ('Croatia', 0.011076571265230286),
 ('Cuba', 0.00947004188672373),
 ('Cyprus', 0.0010821339681852614),
 ('Czech Republic', 0.02097168822090178),
 ('Denmark', 0.015553565537069332),
 ('Djibouti', 0.0008628127696289905),
 ('Dominica', 0.0),
 ('Dominican Republic', 0.007243377483443708),
 ('Ecuador', 0.02427386116940771),
 ('Egypt', 0.006027424782761565),
 ('El Salvador', 0.015443270112446311),
 ('Equatorial Guinea', 0.0),
 ('Eritrea', 0.16435643564356436),
 ('Estonia', 0.06700009436633009),
 ('Ethiopia', None),
 ('Fiji', 0.0),
 ('Finland', 0.11299639583299047),
 ('France', 0.005268360016051791),
 ('Gabon', 0.03880978161735884),
 ('Gambia, The', 0.116600790513834),
 ('Georgia', 0.0),
 ('Germany', 0.023948008443465493),
 ('Ghana', 0.048344635723169826),
 ('Greece', 0.010027019372813766),
 ('Grenada', 0.0),
 ('Guatemala', 0.016144234268703515),
 ('Guinea', 0.0005697611479873187),
 ('Guinea-Bissau', 0.2846728307254623),
 ('Guyana', 0.09205025171578214),
 ('Haiti', 0.006894049346879536),
 ('Honduras', 0.00178746983644651),
 ('Hungary', 0.038166235157575217),
 ('Iceland', 0.02743142144638404),
 ('India', 0.10563390940312317),
 ('Indonesia', 0.0513367141963679),
 ('Iran', 0.07612978626856316),
 ('Iraq', 0.0021720888864500522),
 ('Ireland', 0.02017914434621024),
 ('Israel', 0.02164289227742253),
 ('Italy', 0.024478139661385734),
 ('Jamaica', 0.01477241251961961),
 ('Japan', 0.036846509458551106),
 ('Jordan', 0.006080944122880115),
 ('Kazakhstan', 0.009334370485609512),
 ('Kenya', 0.019726253645851635),
 ('Kiribati', 0.0),
 ('Korea, North', 0.0010796624808982792),
 ('Korea, South', 0.028889806025588115),
 ('Kosovo', 0.0),
 ('Kuwait', 0.0),
 ('Kyrgyzstan', 0.04249195781043894),
 ('Laos', 0.025996533795493933),
 ('Latvia', 0.03759096531671192),
 ('Lebanon', 0.016617790811339198),
 ('Lesotho', 0.0),
 ('Liberia', 0.15623961794019933),
 ('Libya', 0.0),
 ('Liechtenstein', 0.0),
 ('Lithuania', 0.041799617102744095),
 ('Luxembourg', 0.0),
 ('Macedonia', 0.011009318601816538),
 ('Madagascar', 0.009459366509612408),
 ('Malawi', 0.25939625850340137),
 ('Malaysia', 0.0036207961491768015),
 ('Maldives', 0.0),
 ('Mali', 0.016392529032363812),
 ('Malta', 0.0),
 ('Marshall Islands', 0.0),
 ('Mauritania', 0.0),
 ('Mauritius', 0.0049261083743842365),
 ('Mexico', 0.010509556597537481),
 ('Micronesia, Federated States of', 0.0),
 ('Moldova', 0.029187315679061143),
 ('Monaco', 0.0),
 ('Mongolia', 0.006797308883619258),
 ('Montenegro', 0.026761819803746655),
 ('Morocco', 0.000560161326462021),
 ('Mozambique', 0.016531447900506

### Countries with more water than land

In [11]:
%%sql
SELECT name
FROM facts
WHERE area_water > area_land;

 * sqlite:///factbook.db
Done.


[('British Indian Ocean Territory',), ('Virgin Islands',)]

### Countries with the most population growth

In [12]:
%%sql
SELECT name, population_growth
FROM facts
ORDER BY population_growth DESC
LIMIT 5;

 * sqlite:///factbook.db
Done.


[('South Sudan', 4.02),
 ('Malawi', 3.32),
 ('Burundi', 3.28),
 ('Niger', 3.25),
 ('Uganda', 3.24)]

### Countries with more death rate than birth rate

In [13]:
%%sql
SELECT name
FROM facts
WHERE death_rate > birth_rate;

 * sqlite:///factbook.db
Done.


[('Austria',),
 ('Belarus',),
 ('Bosnia and Herzegovina',),
 ('Bulgaria',),
 ('Croatia',),
 ('Czech Republic',),
 ('Estonia',),
 ('Germany',),
 ('Greece',),
 ('Hungary',),
 ('Italy',),
 ('Japan',),
 ('Latvia',),
 ('Lithuania',),
 ('Moldova',),
 ('Monaco',),
 ('Poland',),
 ('Portugal',),
 ('Romania',),
 ('Russia',),
 ('Serbia',),
 ('Slovenia',),
 ('Ukraine',),
 ('Saint Pierre and Miquelon',)]